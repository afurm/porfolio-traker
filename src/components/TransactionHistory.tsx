import React, { useEffect, useState } from 'react';
import axios from 'axios';
import {
  Box,
  Typography,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  useTheme,
} from '@mui/material';

export interface Transaction {
  coinName: string;
  amount: number;
  transactionType: 'buy' | 'sell';
  date: string;
  currentPrice: number; // Price at the time of the transaction
}

export interface TransactionHistoryProps {
  transactions: Transaction[];
}

export interface LatestPrices {
  [key: string]: number | undefined;
}

const TransactionHistory: React.FC<TransactionHistoryProps> = ({ transactions }) => {
  const theme = useTheme();
  const [latestPrices, setLatestPrices] = useState<LatestPrices>({});

  useEffect(() => {
    const fetchLatestPrices = async () => {
      try {
        // Create a unique list of coin names (e.g., ['bitcoin', 'ethereum'])
        const uniqueCoins = Array.from(
          new Set(transactions.map((tx) => tx.coinName.toLowerCase()))
        );

        // Fetch the latest price for all the unique coins from CoinGecko
        const response = await axios.get(
          `https://api.coingecko.com/api/v3/simple/price?ids=${uniqueCoins.join(',')}&vs_currencies=usd`
        );

        // Update the latest prices state
        const fetchedPrices: LatestPrices = {};
        uniqueCoins.forEach((coin) => {
          fetchedPrices[coin] = response.data[coin]?.usd;
        });

        setLatestPrices(fetchedPrices);
      } catch (error) {
        console.error('Error fetching latest prices:', error);
      }
    };

    if (transactions.length > 0) {
      fetchLatestPrices();
    }
  }, [transactions]);

  return (
    <Box
      sx={{
        p: 4,
        backgroundColor: theme.palette.background.default,
        borderRadius: 2,
        boxShadow: 3,
      }}
    >
      <Typography variant="h5" fontWeight="bold" textAlign="center" mb={3}>
        Transaction History
      </Typography>
      {transactions.length === 0 ? (
        <Typography textAlign="center" variant="body1" color="textSecondary">
          No transactions found
        </Typography>
      ) : (
        <TableContainer
          component={Paper}
          elevation={3}
          sx={{ maxWidth: 800, mx: 'auto', borderRadius: 2 }}
        >
          <Table>
            <TableHead>
              <TableRow>
                <TableCell>Coin</TableCell>
                <TableCell>Amount</TableCell>
                <TableCell>Transaction Type</TableCell>
                <TableCell>Transaction Price (USD)</TableCell>
                <TableCell>Current Price (USD)</TableCell>
                <TableCell>Date</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {transactions.map((transaction, index) => {
                // Fetch the latest price for this transaction's coin
                const latestPrice = latestPrices[transaction.coinName.toLowerCase()];

                return (
                  <TableRow
                    key={index}
                    sx={{
                      backgroundColor:
                        transaction.transactionType === 'buy'
                          ? theme.palette.success.lighter
                          : theme.palette.error.lighter,
                      '&:hover': {
                        backgroundColor:
                          transaction.transactionType === 'buy'
                            ? theme.palette.success.light
                            : theme.palette.error.light,
                      },
                    }}
                  >
                    <TableCell>{transaction.coinName}</TableCell>
                    <TableCell>{transaction.amount}</TableCell>
                    <TableCell>
                      <Typography
                        color={
                          transaction.transactionType === 'buy' ? 'success.main' : 'error.main'
                        }
                      >
                        {transaction.transactionType.charAt(0).toUpperCase() +
                          transaction.transactionType.slice(1)}
                      </Typography>
                    </TableCell>
                    <TableCell>${transaction.currentPrice.toFixed(2)}</TableCell>
                    <TableCell>
                      {latestPrice !== undefined ? `$${latestPrice.toFixed(2)}` : 'N/A'}
                    </TableCell>
                    <TableCell>{new Date(transaction.date).toLocaleString()}</TableCell>
                  </TableRow>
                );
              })}
            </TableBody>
          </Table>
        </TableContainer>
      )}
    </Box>
  );
};

export default TransactionHistory;
