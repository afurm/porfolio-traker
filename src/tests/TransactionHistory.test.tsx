import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import TransactionHistory, { TransactionHistoryProps } from '@/components/TransactionHistory';

describe('TransactionHistory Component Snapshot Test', () => {
  it('renders correctly with transactions', () => {
    const props: TransactionHistoryProps = {
      transactions: [
        {
          coinName: 'Bitcoin',
          amount: 2,
          transactionType: 'buy',
          date: '2024-10-01',
          currentPrice: 50000.00
        },
        {
          coinName: 'Ethereum',
          amount: 5,
          transactionType: 'sell',
          date: '2024-10-02',
          currentPrice: 3000.00
        },
      ],
    };
    const { asFragment } = render(<TransactionHistory {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly without transactions', () => {
    const props: TransactionHistoryProps = {
      transactions: [],
    };
    const { asFragment } = render(<TransactionHistory {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
