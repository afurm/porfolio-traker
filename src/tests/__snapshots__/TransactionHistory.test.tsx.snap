// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHistory Component Snapshot Test renders correctly with transactions 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-wm8cac"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-1ffs2s2-MuiTypography-root"
    >
      Transaction History
    </h5>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiTableContainer-root css-1kucasm-MuiPaper-root-MuiTableContainer-root"
      style="--Paper-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Coin
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Amount
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Transaction Type
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Transaction Price (USD)
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Current Price (USD)
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1t7lzbz-MuiTableCell-root"
              scope="col"
            >
              Date
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-jhdmo8-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              Bitcoin
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-8lvigo-MuiTypography-root"
              >
                Buy
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              $50000.00
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              N/A
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              10/1/2024, 3:00:00 AM
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1acr16c-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              Ethereum
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              5
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-394iic-MuiTypography-root"
              >
                Sell
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              $3000.00
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              N/A
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-10ukr6t-MuiTableCell-root"
            >
              10/2/2024, 3:00:00 AM
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TransactionHistory Component Snapshot Test renders correctly without transactions 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-wm8cac"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-1ffs2s2-MuiTypography-root"
    >
      Transaction History
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1vfzsi2-MuiTypography-root"
    >
      No transactions found
    </p>
  </div>
</DocumentFragment>
`;
